hash.go

```diff
@@ -164,12 +170,31 @@ func hashWith(salt []byte, name string) string {
 	d.Write(salt)
 	d.Write(opts.Seed)
 	io.WriteString(d, name)
-	sum := nameBase64.EncodeToString(d.Sum(nil))
+	sum := make([]byte, nameBase64.EncodedLen(d.Size()))
+	nameBase64.Encode(sum, d.Sum(nil))
+	sum = sum[:length]
 
-	// TODO: Just make the first letter uppercase or lowercase as needed.
-	// This is also not needed for non-names, like import paths.
+	if !token.IsIdentifier(name) {
+		return string(sum)
+	}
+	// Keep the result a valid identifier which is equally exported or not.
+	if '0' <= sum[0] && sum[0] <= '9' {
```

Comment:
I find this kinda hard to read, I first thought you've made a logical error, I think this is more straightforward:
```go
if  sum[0] >= '0' && sum[0] <= '9' {
```
You could also the unicode.IsDigit, unicode.IsUpper and unicode.IsLower functions here instead.

(Also applies to the other cases)

Reply:
It's a somewhat common way to write "in range" comparisons; it's what IsUpper/IsLower/etc uses in unicode.

I did initially consider using the unicode funcs. But note that they use runes, so it would make this code more complex given that we'd go from using a byte to using a rune. You could assume that its results will always be single bytes, but that's a bit risky.

I could move these comparisons to our own small ascii-based funcs like `isUpper(b byte) bool`, if you think that would make the code clearer.

Reply:
I agree that our own small ascii based funcs might be a bit easier to read

Reply:
Done, please take a look. I also had to keep the digit conversion in for paths, since they are used as package names.